sabnzbd/downloader.py

```diff
@@ -771,49 +747,44 @@ def run(self):
                 continue
 
             if self.recv_threads > 1:
-                for nw, bytes_received, done in self.recv_pool.map(self.__recv, read):
-                    if bytes_received > last_max_chunk_size:
-                        last_max_chunk_size = bytes_received
-                    self.__handle_recv_result(nw, bytes_received, done)
-                if self.bandwidth_limit:
-                    self.__check_speed()
+                # Submit a process_nw task to the pool for every NewWrapper which is readable
+                for selected in read:
+                    if nw := self.read_fds.get(selected):
+                        self.process_tasks.add(self.recv_pool.submit(self.process_nw, nw))
+
+                # Wait for all the tasks to complete
+                concurrent.futures.wait(self.process_tasks)
+
+                # Clear task list
+                self.process_tasks.clear()
             else:
                 for selected in read:
-                    nw, bytes_received, done = self.__recv(selected)
-                    if bytes_received > last_max_chunk_size:
-                        last_max_chunk_size = bytes_received
-                    self.__handle_recv_result(nw, bytes_received, done)
-                    if self.bandwidth_limit and bytes_received:
-                        self.__check_speed()
-
-    def __recv(self, selected):
-        nw = None
+                    if nw := self.read_fds.get(selected):
```

Comment:
Why do you use `get` and not just `self.read_fds[selected]`? Can it be `None`? If so, maybe it needs a comment how that could be possible.

Reply:
It's keeping the same behaviour @puzzledsab added in https://github.com/sabnzbd/sabnzbd/pull/2438/commits/1edbe38859850577aa21d623f206670a76c1ff24 since that was added I presumed they found throwing KeyError was possible - I'm not sure how but presumably a server can be disabled or something when the downloader thread has already selected it?

So I preferred to push that check into the downloader and use `get` which returns None instead of needing to add try/except in the main downloader thread.

Reply:
Perhaps just a comment "# The connection may no longer be available, such as if the server is disabled" or something?

Reply:
I don't think it should be possible. I just did it to close absolutely every possibility of the try not catching anything that could fail inside the thread.

Reply:
It can't happen, it can just use `self.read_fds[selected]`!